### PR TITLE
[IMP]product: product are search using its attrubute value in sale order...

### PR DESCRIPTION
**Pull request by hmo-odoo**
... line and else

Task: https://www.odoo.com/web#id=11003&view_type=form&model=project.task&menu_id=3942&action=327
PAD: https://pad.odoo.com/p/r.kYIbgeEQsB3GLSLq

### CODE REVIEW

File: addons/product/product.py [hunk 1]
```diff
@@ -1051,7 +1051,7 @@ def name_search(self, cr, user, name='', args=None, operator='ilike', context=No
             positive_operators = ['=', 'ilike', '=ilike', 'like', '=like']
             ids = []
             if operator in positive_operators:
-                ids = self.search(cr, user, [('default_code','=',name)]+ args, limit=limit, context=context)
+                ids = self.search(cr, user, ['|', ('default_code','=',name), ('attribute_value_ids.name', operator, name)]+ args, limit=limit, context=context)
```
Comment (ara-odoo) [inline addons/product/product.py]: follow pep-8 many place missing whitespace after `,`

Comment (gan-odoo) [inline addons/product/product.py]: Done !

**Comment by rim-odoo** [comment]
@gan-odoo Could you rebase and fix conflicts? Thanks

**Comment by gan-odoo** [comment]
@rim-odoo conflicts are fix and branch in green on runbot. Thanks

**Comment by rim-odoo** [comment]
@gan-odoo :+1:

File: addons/product/product.py [hunk 2]
```diff
@@ -1069,26 +1069,26 @@ def name_search(self, cr, user, name='', args=None, operator='ilike', context=No
             positive_operators = ['=', 'ilike', '=ilike', 'like', '=like']
             ids = []
             if operator in positive_operators:
-                ids = self.search(cr, user, [('default_code','=',name)]+ args, limit=limit, context=context)
+                ids = self.search(cr, user, ['|', ('default_code', '=', name), ('attribute_value_ids.name', operator, name)] + args, limit=limit, context=context)
```
Comment (odony) [inline addons/product/product.py]: Please don't modify the search by `default_code`, this must be a separate, independent exact match done before everything else.

File: addons/product/product.py [hunk 3]
```diff
@@ -1069,26 +1069,26 @@ def name_search(self, cr, user, name='', args=None, operator='ilike', context=No
...
                 if not ids:
-                    ids = self.search(cr, user, [('barcode','=',name)]+ args, limit=limit, context=context)
+                    ids = self.search(cr, user, [('barcode', '=', name)] + args, limit=limit, context=context)
```
Comment (odony) [inline addons/product/product.py]: Please don't modify the search by `barcode`, this must be another separate, independent exact match done before everything else.

File: addons/product/product.py [hunk 4]
```diff
@@ -1069,26 +1069,26 @@ def name_search(self, cr, user, name='', args=None, operator='ilike', context=No
...
             if not ids and operator not in expression.NEGATIVE_TERM_OPERATORS:
                 # Do not merge the 2 next lines into one single search, SQL search performance would be abysmal
                 # on a database with thousands of matching products, due to the huge merge+unique needed for the
                 # OR operator (and given the fact that the 'name' lookup results come from the ir.translation table
                 # Performing a quick memory merge of ids in Python will give much better performance
-                ids = self.search(cr, user, args + [('default_code', operator, name)], limit=limit, context=context)
+                ids = self.search(cr, user, args + ['|', ('default_code', operator, name), ('attribute_value_ids.name', operator, name)], limit=limit, context=context)
```
Comment (odony) [inline addons/product/product.py]: Don't modify this one either, it's still matching on default_code only.
The only search calls you can modify are the ones searching on `name`, i.e. lines 1084 and 1086. 
And when you modify it, you should search in the display_name of the variants, as going through a `to-many` relationship (`attribute_value_ids`) will be very bad in terms of performance.

As you can see in the comments, performance is critical in the `name_search` of products. You can test your patch in a database with 50k products and see if it is still fast.
Thanks!

File: addons/product/product.py [hunk 5]
```diff
@@ -1069,26 +1069,26 @@ def name_search(self, cr, user, name='', args=None, operator='ilike', context=No
...
                 if not limit or len(ids) < limit:
                     # we may underrun the limit because of dupes in the results, that's fine
                     limit2 = (limit - len(ids)) if limit else False
                     ids += self.search(cr, user, args + [('name', operator, name), ('id', 'not in', ids)], limit=limit2, context=context)
             elif not ids and operator in expression.NEGATIVE_TERM_OPERATORS:
-                ids = self.search(cr, user, args + ['&', ('default_code', operator, name), ('name', operator, name)], limit=limit, context=context)
+                ids = self.search(cr, user, args + ['&', '&', ('default_code', operator, name), ('name', operator, name), '|', ('attribute_value_ids.name', operator, name), ('attribute_value_ids', '=', False)], limit=limit, context=context)
```
Comment (odony) [inline addons/product/product.py]: Here `operator` is a negative one, so the OR expression you added will not do what we want

File: addons/product/product.py [hunk 6]
```diff
@@ -1069,26 +1069,26 @@ def name_search(self, cr, user, name='', args=None, operator='ilike', context=No
...
             if not ids and operator in positive_operators:
                 ptrn = re.compile('(\[(.*?)\])')
                 res = ptrn.search(name)
                 if res:
-                    ids = self.search(cr, user, [('default_code','=', res.group(2))] + args, limit=limit, context=context)
+                    ids = self.search(cr, user, [('default_code', '=', res.group(2))] + args, limit=limit, context=context)
```
Comment (odony) [inline addons/product/product.py]: In general, please avoid cosmetic code changes in code that your patch should not touch, this adds noise in the diff. You can always do a separate PEP8 commit if you think it is important (this is still relatively useless noise)

**Comment by jholze** [comment]
:+1:

File: addons/product/product.py [hunk 7]
```diff
@@ -915,6 +915,13 @@ def _get_price_extra(self, cr, uid, ids, name, args, context=None):
             result[product.id] = price_extra
         return result
 
+    def _display_product_name(self, cr, uid, ids, name, args, context=None):
+        res = {}
+        for product in self.browse(cr, uid, ids, context):
```
Comment (hmo-odoo) [inline addons/product/product.py]: do not change behavior of name_get. here you should call name_get method.

File: addons/product/product.py [hunk 8]
```diff
@@ -1049,9 +1061,20 @@ def name_search(self, cr, user, name='', args=None, operator='ilike', context=No
             positive_operators = ['=', 'ilike', '=ilike', 'like', '=like']
             ids = []
             if operator in positive_operators:
-                ids = self.search(cr, user, [('default_code','=',name)]+ args, limit=limit, context=context)
+                if operator in ('=ilike', '=like'):
+                    operator = operator[1:]
+                search_name = name.split(' ')
+                query = "SELECT id FROM product_product WHERE display_name " + operator + " '%%%s%%'" % (search_name[0])
```
Comment (hmo-odoo) [inline addons/product/product.py]: avoid sql injection. may be you can use {operator}

File: addons/product/product.py (hunk 7 above)
Comment (gan-odoo) [inline addons/product/product.py]: Done

File: addons/product/product.py (hunk 8 above)
Comment (gan-odoo) [inline addons/product/product.py]: Done.

File: addons/product/product.py [hunk 9]
```diff
@@ -1079,9 +1087,29 @@ def name_search(self, cr, user, name='', args=None, operator='ilike', context=No
...
             if operator in positive_operators:
-                ids = self.search(cr, user, [('default_code','=',name)]+ args, limit=limit, context=context)
+                if operator in ('=ilike', '=like'):
+                    operator = operator[1:]
+                search_name = name.split(' ')
+                arg_list = ['%%%s%%' % search_name[0]]
+                query = """SELECT id FROM product_product
+                        WHERE ({display_name} {operator} {name})
+                        """.format(
+                                display_name='display_name',
```
Comment (hmo-odoo) [inline addons/product/product.py]: use unaccent function like https://github.com/odoo/odoo/blob/9.0/openerp/addons/base/res/res_partner.py#L676

**Comment by odony** [comment]
Given the fact that the names of the products, attributes and attribute values are all translatable, how is this going to work if you rely on stored values?

File: addons/product/product.py (hunk 9 above)
Comment (gan-odoo) [inline addons/product/product.py]: Done!

**Comment by gan-odoo** [comment]
@odony  Yes, name of products, attributes and attributes value are translatable fields. When stored value, it is not working properly with other language.
So i don't store a value, but manage all things(name of products, attribute values and reference code) using SQL query.

**Pull request closed without merging**